### PR TITLE
Add Core Time library for dealing with different date formats.

### PR DESCRIPTION
**Pull request by JoshuaGross**
cc @mookerji @mfine

### CODE REVIEW

File: src/Network/Skylark/Core/Time.hs [hunk 1]
```diff
@@ -0,0 +1,104 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE DeriveDataTypeable  #-}
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving  #-}
```
Comment (JoshuaGross) [inline src/Network/Skylark/Core/Time.hs]: FWIW this is the minimal set of pragmas for this library

File: src/Network/Skylark/Core/Time.hs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+instance FromJSON ISO8601 where
+  parseJSON (String s) = untxt s
+  parseJSON _ = mzero
+
+-- ToJSON instances
+instance ToJSON ISO8601 where
+  toJSON = toJSON . txt
+
+-- Parse time.
+parseTime :: forall a m. TimeFormat (Time a) => Monad m => Text -> m (Time a)
```
Comment (mookerji) [inline src/Network/Skylark/Core/Time.hs]: is the `forall` necessary here?

File: src/Network/Skylark/Core/Time.hs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+parseTime t =
+  Time <$> parseTimeM False defaultTimeLocale (untag f) (unpack t)
+  where
+    f :: Tagged (Time a) String
+    f = format
+
+-- From amazonka. Given a tagged time format, untag it, get its formatter and
+--  format the date.
+renderFormattedTime :: forall a. TimeFormat (Time a) => Time a -> String
```
Comment (mookerji) [inline src/Network/Skylark/Core/Time.hs]: A few sanity tests here for serialization/deserialization would be great too.

File: src/Network/Skylark/Core/Time.hs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+data Time :: Format -> * where
+  Time :: UTCTime -> Time a
+    deriving (Data, Typeable, Generic)
+
+deriving instance Eq   (Time a)
+deriving instance Ord  (Time a)
+deriving instance Read (Time a)
+deriving instance Show (Time a)
+
+-- convert :: Time a -> Time b
```
Comment (mookerji) [inline src/Network/Skylark/Core/Time.hs]: ?

File: src/Network/Skylark/Core/Time.hs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+-- Define time formats for formatting dates
+instance TimeFormat ISO8601 where
+  format = Tagged "%FT%T%z"
+
+-- Txt instances
+instance Txt ISO8601 where txt = pack . renderFormattedTime
+
+-- UnTxt instances
+instance UnTxt ISO8601 where untxt = parseTime
```
Comment (mfine) [inline src/Network/Skylark/Core/Time.hs]: move the functions on to the next line under after the where.

File: src/Network/Skylark/Core/Time.hs (hunk 2 above)
Comment (JoshuaGross) [inline src/Network/Skylark/Core/Time.hs]: yup

**Comment by mfine** [comment]
I brought this up offline, but I think it's worth noting that most of this library is currently and for the foreseeable future unnecessary - there's a single time format and type, `ISO8601`, so all the machinery around a generic `Time` is presently unnecessary. So we're vendoring in a lot of complexity we don't end up needing. Something to consider and think about. Our actual requirements right now could be fully satisfied by a simpler `ISO8601` newtype on `UTCTime` with accompanying instances. Yeah, we could potentially have more time formats, right now this feels like YAGNI.

**Comment by JoshuaGross** [comment]
As soon as we need to communicate with any other service that inputs/outputs time in another format, we'll need something like this - I'm a fan of aggressively YAGNI'ing things, but assuming all dates will be in one format seems unreasonable, even in the short term?

**Comment by mfine** [comment]
When / if we do run into a hypothetical other service with different time formats, would there be anything preventing us from moving to a tagged approach?

I'd strongly prefer a simple newtype instance with a few instances for now over rationalizing a tagged approach with a single tag, and switch to a tagged approach down the road when we actually need it.

**Pull request merged**